### PR TITLE
Added Pod Security Policy to the chart

### LINKED ISSUE #450: Add pod security policy to the helm chart
**Describe the request**
Right now we have a pod security policy enabled in our k8s clusters that prevent the creation of certain volumes and allow only certain capabilities, etc.. this is a blocker to deploy secrets-store-csi-driver properly so we had to create the following pod security policy manually.

```
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  name: secrets-store
spec:
  seLinux:
    rule: RunAsAny
  privileged: true
  volumes:
    - hostPath
    - secret
  hostNetwork: true
  hostPorts:
  - min: 0
    max: 65535
  fsGroup:
    rule: RunAsAny
  runAsUser:
    rule: RunAsAny
  supplementalGroups:
    rule: RunAsAny
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: psp-secrets-store-role
  namespace: NAME
rules:
- apiGroups: ['extensions']
  resources: ['podsecuritypolicies']
  verbs:     ['use']
  resourceNames:
  - secrets-store
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: psp-secrets-store-role-binding
  namespace: NAME
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: psp-secrets-store-role
subjects:
- kind: ServiceAccount
  name: secrets-store-csi-driver
  namespace: NAME
```

**Explain why AAD Pod Identity needs it**
To deploy secrets-store-csi-driver successfully in k8s cluster with a restricted pod-security policy enabled.

**Describe the solution you'd like**
Merge the pod security policy mentioned above in the software helm chart, and keep it updated with the new software releases.

**Describe alternatives you've considered**
the other alternative is to create the pod security policy manually, which is not good if more capabilities are needed with new sw releases and we have always to update the yaml manually

**Environment:**

- Secrets Store CSI Driver version: (use the image tag): latest
- Kubernetes version: (use `kubectl version`): v1.18.8

### PR DESCRIPTION
**Pull request by pierluigilenoci**
**What this PR does / why we need it**:

We want to deploy secrets-store-csi-driver successfully in k8s cluster with a restricted pod-security policy enabled.

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #450 

**Special notes for your reviewer**:

This is inspired by @mo-saeed code from #450 
@aramase feel free to change the name of the role, policy and rolebinding.

### CODE REVIEW

File: manifest_staging/charts/secrets-store-csi-driver/README.md [hunk 1]
```diff
@@ -78,6 +78,7 @@ The following table lists the configurable parameters of the csi-secrets-store-p
 | `livenessProbe.port`                    | Liveness probe port                                                                                                               | `9808`                                              |
 | `livenessProbe.logLevel`                | Liveness probe container logging verbosity level                                                                                  | `2`                                                 |
 | `rbac.install`                          | Install default rbac roles and bindings                                                                                           | true                                                |
+| `rbac.pspEnabled`                       | If `true`, create and use a restricted pod security policy for Secrets Store CSI Driver pod(s)                                                                                       | true                                                                                             |
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/README.md:81]: Default should be false as that's the current behavior

File: manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{{- if .Values.rbac.pspEnabled }}
+apiVersion: policy/v1beta1
+kind: PodSecurityPolicy
+metadata:
+  name: secrets-store-csi-driver-psp
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml:5]: If there is no privileged psp, wouldn't the pod creation request fail after the `restricted` psp is evaluated first?

File: manifest_staging/charts/secrets-store-csi-driver/templates/role-psp.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+kind: Role
+metadata:
+  name: secrets-store-csi-driver-role-psp
+  namespace: {{ .Release.Namespace }}
+rules:
+  - apiGroups: [ 'extensions' ]
+    resources: [ 'podsecuritypolicies' ]
+    verbs: [ 'use' ]
+    resourceNames:
+      - secrets-store
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/role-psp.yaml:12]:
```suggestion
      - secrets-store-csi-driver-psp
```

This should match the psp name from above

File: manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml (hunk 2 above)
Comment (pierluigilenoci) [reply manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml:5]: If one enables the PSP it is assumed that the cluster uses them and therefore at least one unprivileged policy should be there. Cert-Manager also uses this approach.
https://github.com/jetstack/cert-manager/blob/master/deploy/charts/cert-manager/templates/psp.yaml

File: manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+spec:
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml:6]: We should also include the labels here similar to https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml#L7

**Review by aramase — COMMENTED** [review]
I think we should be able to add the pod security policy manifests to https://github.com/kubernetes-sigs/secrets-store-csi-driver/tree/master/config/rbac and psp for use to existing cluster role using `kubebuilder` similar to https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/controllers/secretproviderclasspodstatus_controller.go#L210-L214.

File: manifest_staging/charts/secrets-store-csi-driver/templates/role-psp.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+{{- if .Values.rbac.pspEnabled }}
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: secrets-store-csi-driver-role-psp
+  namespace: {{ .Release.Namespace }}
+rules:
+  - apiGroups: [ 'extensions' ]
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/role-psp.yaml:8]:
```suggestion
  - apiGroups: [ 'policies' ]
```

Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/role-psp.yaml:8]:
```suggestion
  - apiGroups: [ 'policy' ]
```

Comment (pierluigilenoci) [reply manifest_staging/charts/secrets-store-csi-driver/templates/role-psp.yaml]: fixed

File: manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml (hunk 4 above)
Comment (pierluigilenoci) [reply manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml:7]: Done

**Comment by pierluigilenoci** [comment]
@aramase let me know what to do and I do. 😄

**Comment by pierluigilenoci** [comment]
@aramase I improved the chart, let me know what you think about it.

File: manifest_staging/charts/secrets-store-csi-driver/templates/_helpers.tpl [hunk 6]
```diff
@@ -33,6 +33,10 @@ labels:
   helm.sh/chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
 {{- end -}}
 
+{{- define "sscd-psp.fullname" -}}
+{{- printf "%s-psp-mic" (include "sscd.name" .) -}}
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/_helpers.tpl:37]: The name here shouldn't have `mic` suffix

File: manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml [hunk 7]
```diff
@@ -0,0 +1,24 @@
+{{- if .Values.rbac.pspEnabled }}
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml:1]:
```yaml
{{- if .Values.rbac.pspEnabled }}
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  name: {{ template "sscd-psp.fullname" . }}
{{ include "sscd.labels" . | indent 2 }}
spec:
  seLinux:
    rule: RunAsAny
  privileged: true
  volumes:
    - hostPath
    - secret
  hostNetwork: false
  hostPorts:
    - min: 0
      max: 65535
  fsGroup:
    ranges:
    - max: 65535
      min: 1
    rule: MustRunAs
  runAsUser:
    rule: RunAsAny
  supplementalGroups:
    ranges:
    - max: 65535
      min: 1
    rule: MustRunAs
{{- end }}
```

**Review by aramase — CHANGES_REQUESTED** [review]

File: manifest_staging/charts/secrets-store-csi-driver/templates/_helpers.tpl (hunk 6 above)
Comment (pierluigilenoci) [reply manifest_staging/charts/secrets-store-csi-driver/templates/_helpers.tpl:37]: done

File: manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml (hunk 7 above)
Comment (pierluigilenoci) [reply manifest_staging/charts/secrets-store-csi-driver/templates/podsecuritypolicy.yaml:1]: done

**Comment by pierluigilenoci** [comment]
@aramase Thank you so much for your help and your patience. ❤️

**Review by aramase — COMMENTED** [review]
/ok-to-test

**Comment by aramase** [comment]
@pierluigilenoci Could you squash the commits?

**Comment by pierluigilenoci** [comment]
@aramase done

**Review by aramase — COMMENTED** [review]
/lgtm

**Comment by pierluigilenoci** [comment]
So we have to wait for @ritazh for `approve`?

**Review by aramase — COMMENTED** [review]
/approve

**Comment by NoPrivilege** [comment]
Hello everybody,

I assume you added this PR for the same reasons i am writing theses lines.
Nevertheless, let me expose you the issue we are encountering.  We are currently using the "secrets store csi driver" in my organization. Our problem is it runs as a privileged pod.

QUESTION : Is it possible to run this "csi driver" as non privileged pod and still be able to perform the same tasks? 

IF YES :  i would be grateful to have some documentation.

And if NOT, is there a workaround ? is this new PR able to mitigate the risk of running a privileged pod?

Thank you in advance for your reply

**Pull request merged**